src/term/mod.rs

```diff
@@ -200,6 +200,65 @@ pub use self::mode::TermMode;
 
 pub const TAB_SPACES: usize = 8;
 
+#[derive(Clone, Copy)]
+enum Charset {
+    Ascii,
+    /// A sparse charset only has a few characters in it and is best mapped with
+    /// a match statement
+    Sparse(StandardCharset),
+    /// A dense charset has most of the 96 possible characters in it and may be
+    /// better handler by a lookup table
+    Dense(StandardCharset),
+    /// A user defined charset handled by some more complicated mechanism
+    User,
+}
+
+impl Charset {
+    /// Switch/Map character to the active charset. Ascii is the common case and
+    /// for that we want to do as little as possible.
+    #[inline]
+    fn switch(&self, c: char) -> char {
+        match *self {
+            Charset::Ascii => c,
+            Charset::Sparse(set) => match set {
+                StandardCharset::LineDrawing =>
+                    match c {
```

Comment:
Yeah, I would be interested to see how the two approaches compare (at some point). If the Rust compiler/LLVM is clever enough they should be the same or the match statement should be faster.